src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -319,6 +321,8 @@ async def mypy_typecheck_partition(
     env = {
         "PEX_EXTRA_SYS_PATH": ":".join(all_used_source_roots),
         "MYPYPATH": ":".join(all_used_source_roots),
+        "MYPY_FORCE_COLOR": "1" if global_options.colors else "0",
+        "MYPY_FORCE_TERMINAL_WIDTH": str(shutil.get_terminal_size().columns),
```

Comment:
I'm not clear if this is appropriate in two ways:

- will this ever be running in the pantsd, where it may not have access to the real tty (and/or have an outdated understanding of the size)?
- I note that there's a function in `docutil.py`, but that seems to be targeted at help text? Should I be reusing it here? https://github.com/pantsbuild/pants/blob/073062a0d567edfd19f3f0fdc753cc15d4645b76/src/python/pants/util/docutil.py#L15-L16

Reply:
`pantsd` propagates the TTY, so this should work as is.

But there is another thing to consider here, which is that including this information makes the cache key of the process dependent on your terminal width and the value of the `--color` flag, which isn't ideal.

For color in particular, we apparently do [the same thing for `pytest`](https://github.com/pantsbuild/pants/blob/29d7359cc5097ff907996dee5a07f989f586f4fa/src/python/pants/backend/python/goals/pytest_runner.py#L283)... which is also not great. Instead, what would likely be better to do for `color` in particular (for both `pytest` and `mypy`) would be to always request it in the consuming process, but then strip it (via e.g. https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python) from stdout/stderr afterwards before creating the `CheckResult`.

Terminal width is a separate issue: it cannot be adjusted post-facto. One option would be to hardcode it to e.g. `100` based on a global option (so that folks could fiddle with it in case the default wasn't good). Another option (which would be more magical but maybe totally fine) would be to round the terminal width down to the nearest 10 or 20 (or to 80 as a minimum) so that everyone with a terminal width within 10 or 20 characters of one another could still hit the cache.

Reply:
Hm, that's a real bummer on terminal width...it's going to be a barrier to "desktop caching" working well, where remote cache can be shared amongst several users.

I like your suggestion for post-processing colors for both MyPy and Pytest.

Reply:
cc @Eric-Arellano for any thoughts on the safety of stripping color information.

Reply:
> Hm, that's a real bummer on terminal width...it's going to be a barrier to "desktop caching" working well, where remote cache can be shared amongst several users.

With a sufficiently granular rounding threshold (probably more like the nearest 40 or maybe even 80 characters), the hitrate would still be good. It's just annoyingly magical.

Reply:
Still, even for local caching I'm not sure about it. If I expand my terminal because I went from 13" laptop to ultrawide monitor, I would not expect MyPy to not be invalidated. I actually make that change a lot!

Perhaps we don't want to land the terminal width change yet?

Reply:
Good point, yea. So maybe globally hardcoding it really is best. It's likely easiest anyway.

Reply:
It sounds like there is no alternative fix where we get MyPy to believe it's a TTY, then? Rather than us hardcoding the args like this.

Reply:
To be clear: the existing diff in the PR doesn't make mypy "believe it's a TTY" either: it just tells it what width to use when rendering. But yea, I can't imagine a way to truly make it render to the TTY without making it an `InteractiveProcess` (...which would mean no caching).

Reply:
Thanks for the discussion.

I'll make the change to mypy to set `MYPY_FORCE_COLOR=1` always, and then call `colors.strip_color` (but let me know if you'd prefer pants to write its own form) or not, depending on the colour option. To keep this PR focused (since I have limited time to work on it 😄 ), I won't touch the pytest code here.

For terminal width, one option might be to hardcode to 'infinity' (9999999 or something), essentially disabling mypy's hard-wrapping entirely by default. This won't work for pytest, but should work for mypy: I'll do some experiments to confirm. 

1. the terminal can soft-wrap diagnostic text just fine
2. source code typically is typically natively hard-wrapped to some manageable limit already (80, 100, 120 characters)

IIRC, this is more along the lines of what other compilers do anyway (at least 1), e.g. rustc in a 60 char terminal:

![image](https://user-images.githubusercontent.com/1203825/184734277-4b9c2931-b3ce-4db7-bbc8-4ffa42990dc8.png)

Reply:
I think the `infinity` approach is really promising, as long as the terminal indeed will do soft wrapping. Imo, that's preferable to a global option for MyPy, which is annoying to ask the user to set and still breaks caching. Thanks!

Sounds good re: Pytest, I'll open a ticket.

Reply:
I've updated this along those lines. I think the output is acceptable, even for very wide code (200+ cols) in a narrow terminal:

![image](https://user-images.githubusercontent.com/1203825/184791961-83bd7346-b88d-4469-afc9-f84bfaf757da.png)


A few notes for this new version:

1. I've put the color stripping into `CheckResult.from_fallible_process_result`, so it can be used for other checks too. I imagine this could be used in `LintResult` and even `TestResult` but it kinda seems like there's deeper sharing between those types so I haven't tried to merge them here.
2. I've had to set `TERM` too to get colors (mypy uses `curses` to do it 'properly' rather than the rather more convenient approach of hardcoding the escapes ala `ansicolors`). From skimming https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/8 and https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/11, it seemed like `TERM=xterm` was a reasonable lowest common denominator for terminal emulates used in practice (especially for the simple escape sequences mypy uses https://github.com/python/mypy/blob/9a9bc3b807bb11986122dd3896e6eb3b61e2ccf8/mypy/util.py#L626-L633), but I don't have much insight here, so please let me know if there's a preferred choice.
3. I still haven't written any tests for the actual mypy invocations, assuming the code looks reasonable, I'll add something to `rules_integration_test.py`

Reply:
> I still haven't written any tests for the actual mypy invocations, assuming the code looks reasonable, I can add something to rules_integration_test.py

Ah, looks like I have to update it anyway. It'll probably have to wait to tomorrow on my end, though.

Reply:
> it seemed like TERM=xterm was a reasonable lowest common denominator for terminal emulates used in practice (especially for the simple escape sequences mypy uses https://github.com/python/mypy/blob/9a9bc3b807bb11986122dd3896e6eb3b61e2ccf8/mypy/util.py#L626-L633), but I don't have much insight here, so please let me know if there's a preferred choice

Ah, unfortunately this didn't work: for xterm, terminfo (at least on my macOS system) uses a more complicated escape sequence for `sgr0` (reset) `\x1b(B\x1b[m`, instead of just some variant of the normal `\x1b[0m`. `colors.strip_color` doesn't handle the `\x1b(B` escape, nor do any of the (correct) regexps in the StackOverflow discussion above. 

Instead of trying to patch holes like this or pull in a library like https://github.com/selectel/pyte to do it fully properly, I've just switched to running mypy with a simpler terminal emulator specified (`TERM=linux`). 

```shell
TERM=xterm python -c 'import curses; curses.setupterm(fd=1); print(repr(curses.tigetstr("sgr0").decode()))'
# '\x1b(B\x1b[m'

TERM=linux python -c 'import curses; curses.setupterm(fd=1); print(repr(curses.tigetstr("sgr0").decode()))'
# '\x1b[0;10m'
```

I've also added a basic smoke test since this took me so long to actually get working properly 😅